### PR TITLE
16 pawn role management

### LINKED ISSUE #16: Pawn Role Management
## Description
As a player I want to be able to assign my pawns to specific roles in order to have greater control over my colony.

### Functional
- [x] There exists different roles: lumberjack, farmer, guard, miner, fisher, builder
- [x] There is a mechanism for assigning idle pawns to roles when requested
- [x] There is a menu for assigning a number of pawns to each role. The number can not exceed the total number of pawns
- [x] A pawn can switch roles when the player sets a quota for a role

## Dependencies
- #9

### PR DESCRIPTION
**Pull request by martinjonsson01**
Implements managing of roles by the player, along with some tweaks to pawn behaviour.

There is currently only a class diagram included for `com.thebois.listeners`, the rest will be added shortly.

Closes #16

### CODE REVIEW

File: src/main/java/com/thebois/models/beings/AbstractBeing.java [hunk 1]
```diff
@@ -78,18 +96,27 @@ protected void move() {
...
+        return Objects.equals(getDestination(), that.getDestination()) && Objects.equals(
+            currentPosition,
+            that.currentPosition) && Objects.equals(getRole(), that.getRole());
     }
 
-    @Override
-    public void update() {
-        move();
+    protected Position getDestination() {
+        return this.destination;
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/AbstractBeing.java:115]: Deep clone?

File: src/main/java/com/thebois/models/beings/AbstractBeingGroup.java [hunk 2]
```diff
@@ -8,50 +8,37 @@
...
-     *
-     * @param being IBeing to be added.
-     */
-    protected void addBeing(IBeing being) {
-        beings.add(being);
+    @Override
+    public Collection<IBeing> getBeings() {
+        final Collection<IBeing> clonedBeings = new ArrayList<>(beings.size());
+        clonedBeings.addAll(beings);
+        return clonedBeings;
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/AbstractBeingGroup.java:33]: 👍

File: src/main/java/com/thebois/models/beings/AbstractBeingGroup.java [hunk 3]
```diff
@@ -8,50 +8,37 @@
...
 
-    private void createBeing(int numberOfBeings) {
-        for (int i = 0; i < numberOfBeings; i++) {
-            addBeing(new Pawn());
-        }
+    protected void setBeings(final Collection<IBeing> beings) {
+        this.beings = beings;
+    }
+
+    protected Collection<IBeing> getBeingsByReference() {
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/AbstractBeingGroup.java:40]: Do we need this?

File: src/main/java/com/thebois/models/beings/Colony.java [hunk 4]
```diff
@@ -1,25 +1,133 @@
...
+
+    private boolean isIdle(final IBeing being) {
+        return being.getRole() == null;
+    }
+
+    private Collection<IBeing> findBeingsWithRole(final AbstractRole role) {
+        return getBeingsByReference().stream()
+                                     .filter(being -> being.getRole() != null)
+                                     .filter(being -> role.equals(being.getRole()))
+                                     .collect(Collectors.toList());
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/Colony.java:130]: This could be ColonyRoleManager. Right now its borderline OK imo but if we need more in the future we should move these

File: src/main/java/com/thebois/models/beings/IBeing.java [hunk 5]
```diff
@@ -10,10 +11,24 @@
     /**
      * Gets the current location.
      *
-     * @return The current location.
+     * @return The current location
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/IBeing.java:14]: What does the codestyle want? I guess either is ok?

File: src/main/java/com/thebois/models/beings/Pawn.java [hunk 6]
```diff
@@ -1,20 +1,20 @@
...
  */
 public class Pawn extends AbstractBeing {
 
-    /**
-     * Creates a Pawn.
-     */
-    public Pawn() {
-    }
+    /* Temporary hard-coded world size. Should be removed when pathfinding is implemented. */
+    private static final int WORLD_SIZE = 50;
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/Pawn.java:13]: Can world size be a static in `World` that we assign in its constructor? (Like if its possible in JAVA) If so we can remove this field and have WORLD.size

What do you think?

Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/beings/Pawn.java:13]: Feels a bit random that the pawn has a hardcoded value of the world, since we can change the value at  any time, and then we have to update it here aswell

File: src/main/java/com/thebois/models/beings/roles/RoleFactory.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+     * @return The role with the specified type
+     *
+     * @throws java.util.NoSuchElementException If there is no such role
+     */
+    public static AbstractRole fromType(final RoleType roleType) {
+        return all()
+            .stream()
+            .filter(role -> role.getType().equals(roleType))
+            .findFirst()
+            .orElseThrow();
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/beings/roles/RoleFactory.java:28]: Seen this at multiple places, what does it do?

File: src/main/java/com/thebois/models/world/World.java [hunk 8]
```diff
@@ -1,26 +1,33 @@
...
 
 /**
  * World creates a matrix and keeps track of all the structures and resources in the game world.
  */
 public class World implements IFinder {
 
-    private ITerrain[][] terrainMatrix;
-    private IBeingGroup colony;
+    private final ITerrain[][] terrainMatrix;
+    private final int beingCount;
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/models/world/World.java:19]: is this being count or pawn count?

File: src/main/java/com/thebois/utils/MatrixUtils.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+     *
+     * @param matrix     The matrix to operate on
+     * @param action     The action to perform for each element
+     * @param <TElement> The type of the matrix elements
+     */
+    public static <TElement> void forEachElement(final TElement[][] matrix,
+                                                 final Consumer<TElement> action) {
+        for (final TElement[] row : matrix) {
+            for (final TElement element : row) {
+                action.accept(element);
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/utils/MatrixUtils.java:25]: Accept just preforms the action on the element?

File: src/main/java/com/thebois/views/ColonyView.java [hunk 10]
```diff
@@ -1,42 +1,110 @@
...
+        ROLE_COLORS = new HashMap<>();
+        ROLE_COLORS.put(RoleType.BUILDER, Color.LIGHT_GRAY);
+        ROLE_COLORS.put(RoleType.FARMER, Color.GREEN);
+        ROLE_COLORS.put(RoleType.FISHER, Color.BLUE);
+        ROLE_COLORS.put(RoleType.GUARD, Color.BLACK);
+        ROLE_COLORS.put(RoleType.LUMBERJACK, Color.BROWN);
+        ROLE_COLORS.put(RoleType.MINER, Color.GRAY);
+    }
+
+    private final Color beingColor = Color.WHITE;
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/views/ColonyView.java:39]: `defaultPawnColor` or if we have a default role `laborerColor`

File: src/main/java/com/thebois/views/ColonyView.java [hunk 11]
```diff
@@ -1,42 +1,110 @@
...
-            for (IBeing pawn : colony.getBeings()) {
-                posX = pawn.getPosition().getPosX();
-                posY = pawn.getPosition().getPosY();
+            for (final IBeing pawn : colony.getBeings()) {
+                posX = pawn.getPosition().getPosX() * tileSize;
+                posY = pawn.getPosition().getPosY() * tileSize;
 
-                batch.setColor(Color.WHITE);
-                batch.circle(posX, posY, radius);
+                if (pawn.getRole() != null) {
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/views/ColonyView.java:89]: See related comment to avoid null

File: src/main/java/com/thebois/views/ColonyView.java [hunk 12]
```diff
@@ -1,42 +1,110 @@
...
+                           offsetY + posY,
+                           radius * 2,
+                           radius * 2);
             }
         }
     }
 
+    @Override
+    public void dispose() {
+        beingTexture.dispose();
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/views/ColonyView.java:107]: Should the view handle the disposing or gamescreen?

File: src/main/java/com/thebois/views/IActorView.java [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+ * Represents a view that does not draw itself, but is rendered using actors instead.
+ */
+public interface IActorView {
+
+    /**
+     * Gets the root element of the view.
+     *
+     * @return The root element containing all other parts of the view.
+     */
+    Actor getRoot();
```
Comment (jonathan-lindqvist) [inline src/main/java/com/thebois/views/IActorView.java:15]: Seems like a code smell? Generally a child should not know about its parent, maybe I am misunderstanding the implementation/use-case

**Review by jonathan-lindqvist — CHANGES_REQUESTED** [review]
Its hard to review 2k+ lines, but it seems good, small changes required and some explanations. And test coverage %?

This feels like at least two user stories, maybe even three. You refactored the whole view part and implemented Role (and wrote tests for all classes that were not implemented).

This should have been made a separate user story so, lets say `refactor rendering`, so that part can be completed first and other team members that also need that change can get that faster. And then the implementation of the Roles user story is not stopping that refactor from reaching master sooner.

Also when making these large PRs its a lot easier for "bad code"/code smells and bugs to make it past the review process. Not to sound harsh and I know you understand but in a work environment this would probably be denied without second thought, if it didn't need to be this large of course.

But really nice to get more tests! 👍

File: colony-manager.iml [hunk 14]
```diff
@@ -25,10 +25,11 @@
     <orderEntry type="library" scope="TEST" name="Maven: org.junit.jupiter:junit-jupiter-engine:5.7.0" level="project" />
     <orderEntry type="library" scope="TEST" name="Maven: org.junit.platform:junit-platform-engine:1.7.0" level="project" />
     <orderEntry type="library" scope="TEST" name="Maven: org.assertj:assertj-core:3.20.2" level="project" />
-    <orderEntry type="library" scope="TEST" name="Maven: org.mockito:mockito-core:3.12.4" level="project" />
-    <orderEntry type="library" scope="TEST" name="Maven: net.bytebuddy:byte-buddy:1.11.13" level="project" />
-    <orderEntry type="library" scope="TEST" name="Maven: net.bytebuddy:byte-buddy-agent:1.11.13" level="project" />
-    <orderEntry type="library" scope="TEST" name="Maven: org.objenesis:objenesis:3.2" level="project" />
+    <orderEntry type="library" name="Maven: org.mockito:mockito-core:3.12.4" level="project" />
```
[the remark this answers is not in the document]
Comment (martinjonsson01) [reply colony-manager.iml]: Nope, it's fixed now!

File: src/main/java/com/thebois/ColonyManagement.java [hunk 15]
```diff
@@ -1,82 +1,116 @@
...
-    private final int worldSize = 50;
-    private final float viewportWidth = 1000f;
-    private final float viewportHeight = 1000f;
-    // Models
-    private World world;
-    // Views
-    private WorldView worldView;
-    private ColonyView colonyView;
+    private TextureAtlas atlas;
+    private Skin skin;
```
[the remark this answers is not in the document]
Comment (martinjonsson01) [reply src/main/java/com/thebois/ColonyManagement.java:40]: I could call `atlas` `skinAtlas` instead, but I don't think that would clear anything up. Short of defining what a `Skin` and texture atlas actually is, I can't think of any better names. 

`skin` could maybe be called `uiSkin` to clarify that it's for the UI?

File: src/main/java/com/thebois/models/beings/AbstractBeing.java [hunk 16]
```diff
@@ -1,24 +1,22 @@
...
+import java.util.Objects;
 
-import com.thebois.models.IDeepClonable;
 import com.thebois.models.Position;
+import com.thebois.models.beings.roles.AbstractRole;
 
 /**
  * An abstract implementation of IBeing.
  */
-public abstract class AbstractBeing implements IBeing, IDeepClonable<IBeing> {
```
[the remark this answers is not in the document]
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/AbstractBeing.java:11]: We don't want `IDeepClonable<IBeing>` because `AbstractBeing` is a base class for an Entity, and they shouldn't be immutable and therefore there is no need for it being clonable.

File: src/main/java/com/thebois/models/beings/AbstractBeing.java [hunk 17]
```diff
@@ -34,11 +32,32 @@ public AbstractBeing() {
...
      */
-    public AbstractBeing(Position currentPosition, Position destination) {
+    public AbstractBeing(final Position currentPosition, final Position destination) {
         this.currentPosition = currentPosition;
         this.destination = destination;
     }
 
+    @Override
+    public Position getPosition() {
+        return this.currentPosition;
```
[the remark this answers is not in the document]
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/AbstractBeing.java:42]: Good catch, I'll look into it.

**Comment by martinjonsson01** [comment]
@jonathan-lindqvist yeah that's fair, it's a huge PR. Part of the problem is that I needed to pull in a lot of changes from the other branches, which meant I had to modify everything I pulled in to match the new systems.

The other part is that implementing the InfoView and the new way of rendering views actually *is* a part of the user story #16. Having user stories "for the programmers" like "As a programmer I want to refactor the view rendering so that there can be UI elements." isn't correct iirc. Since the user stories are vertical slices of everything that needs to be implemented to finish the story, the rendering refactor is supposed to be baked into the user story.

On the other hand, I agree that it's still a good idea to split out these "user story agnostic"-implementations into separate PRs. We could do those in branches separate from the user story branch that actually needs them, to get the benefits you mentioned. Then the problem would be linking branches to their corresponding user stories, but the benefits probably outweigh the downsides. 

Looking at it from a different perspective though, no matter how much we split up this PR into multiple different PRs from multiple branches, you'd still have to review all of it. You'd just be forced to split it up over a few days. For this PR, you can still review it in parts and come back to it over the span of a few days until you're done. Doesn't necessarily have to be completed in a single session, imo.

File: src/main/java/com/thebois/models/beings/AbstractBeing.java [hunk 18]
```diff
@@ -34,11 +32,32 @@ public AbstractBeing() {
...
+    }
+
+    @Override
+    public AbstractRole getRole() {
+        if (role == null) return null;
```
[the remark this answers is not in the document]
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/AbstractBeing.java:47]: Great idea, I was thinking of implementing an `IdleRole` for just that reason. I decided against it though to make the PR shorter, but I suppose we're way past that 😛

File: src/main/java/com/thebois/models/beings/AbstractBeing.java [hunk 19]
```diff
@@ -78,18 +96,27 @@ protected void move() {
...
+        return Objects.hash(getDestination(), currentPosition, getRole());
     }
 
     @Override
-    public Position getPosition() {
-        return this.currentPosition;
+    public boolean equals(final Object o) {
+        if (this == o) return true;
+        if (!(o instanceof AbstractBeing)) return false;
+        final AbstractBeing that = (AbstractBeing) o;
```
[the remark this answers is not in the document]
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/AbstractBeing.java:108]: This was actually the autogenerated name when auto-implementing `equals` and `getHashCode`. I liked the parallels to `this`. I'm comparing `this` to `that`, you know?

File: src/main/java/com/thebois/models/beings/AbstractBeing.java (hunk 1 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/AbstractBeing.java:115]: Good catch!

File: src/main/java/com/thebois/models/beings/AbstractBeingGroup.java (hunk 3 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/AbstractBeingGroup.java:40]: Not anymore, you're right! I had to create that method when the other `getBeings` returned deep clones of the real beings. `Colony` needs access to the actual `IBeing` instances, so I had to create this protected method. Forgot to remove it.

File: src/main/java/com/thebois/models/beings/Colony.java (hunk 4 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/Colony.java:130]: That's what I thought as well. My original implementation had `Colony` use a `RoleAllocator`. I encountered problems though, because the `RoleAllocator` only operates directly on the `IBeing` instances inside `Colony`. That's a code smell, which points towards that it is `Colony`s responsibility to handle the role management.

I actually asked about this in our Slack and you guys said to just put it in `Colony` 😛

File: src/main/java/com/thebois/models/beings/IBeing.java (hunk 5 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/IBeing.java:14]: Yeah, that's something I've noticed that we all use different standards for.

I see them as *sentence fragments* that don't need to be a complete sentence in and of themselves, as opposed to the actual javadoc comment for the method itself. But I've seen code with periods as well. I've also seen some without capital first letter and without period. We should probably pick one and stick to it.

I'll look into adding this to the checkstyle!

File: src/main/java/com/thebois/models/beings/Pawn.java (hunk 6 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/Pawn.java:13]: "/* Temporary hard-coded world size. Should be removed when pathfinding is implemented. */" 

It was either that, or send a reference for the `World` instance to `Pawn` which is even worse. 

That's a nice idea, we could place it as a `public static` constant in `World`. Problem with that is that is global data, which is still a huge code smell. 

My reasoning was that no matter how we implement this right now, it's too much work for something that'll be removed right after this PR is merged and I begin work on #13 😛

File: src/main/java/com/thebois/models/beings/roles/RoleFactory.java (hunk 7 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/roles/RoleFactory.java:28]: I recommend reading about the [Java Stream API](https://www.baeldung.com/java-8-streams#pipeline), it's hugely useful when working with collections of objects.

To break down what is happening here line by line:

`all()` just gets all roles.

`.stream()` just converts the `Collection<AbstractRole>` into a temporary `Stream<AbstractRole>` that can be operated on.

`.filter(role -> role.getType().equals(roleType))` does exactly what it says it does, it filters. Based on what? Based on the lambda given as the argument. In this case `role-> role.getType().equals(roleType)`. The lambda is called for every element in the collection and returns whether or not to keep the elements or exclude them.

`.findFirst()` gets the first element of the resulting filtered `Stream`, but since there might be no elements in it, it returns an `Optional<AbstractRole>` which has a method called `.orElseThrow()` which throws an error if it does not contain any value.

Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/roles/RoleFactory.java:28]: So in layman's terms, it gets all roles, filters to only keep roles with a type equal to `roleType`, tries to get the first result, and throws an error if there were no roles that matched the filter.

File: src/main/java/com/thebois/models/world/World.java (hunk 8 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/world/World.java:19]: It used to be more general, to generate any amount of beings, but it's since changed. 

I'll update it to `pawnCount`.

File: src/main/java/com/thebois/utils/MatrixUtils.java (hunk 9 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/utils/MatrixUtils.java:25]: Exactly!

File: src/main/java/com/thebois/views/ColonyView.java (hunk 10 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/views/ColonyView.java:39]: Good idea, once I've added a role for idle pawns I'll change this too.

File: src/main/java/com/thebois/views/ColonyView.java (hunk 12 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/views/ColonyView.java:107]: Since the texture here is created in, and only ever used in, `ColonyView`, it should be `ColonyView` that is responsible for disposes of it, no? 

`GameScreen` on the other hand, is responsible for disposing of the `Stage` it contains. The `Stage`, in turn, is responsible for disposing the `Widgets` inside of it (that is, `InfoView` and `GameView`), who are then responsible for disposing of their `IView`s, who are then in turn responsible for disposing of any textures they use. 

I like the chain-pattern here, it's kind of elegant, don't you think?

File: src/main/java/com/thebois/views/IActorView.java (hunk 13 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/views/IActorView.java:15]: It's not a child knowing about its parent, it's more a parent knowing about its children. 

The `IActorView` interface is implemented by `RoleView`, which is actually just made up out of widgets from LibGDX. These widgets are all placed in a *root element* in `RoleView`, and in order to render `RoleView`, `GameScreen` needs to add this root element to the `Stage`. So it's actually the parent, `GameScreen`, getting its child through the `getRoot()` method.

Maybe the name is a bit confusing, do you have any ideas for different names?

**Comment by martinjonsson01** [comment]
As for coverage, it's currently at 85% lines, and 51% branches:
image

com.thebois.world is at 50% branch coverage:
image

com.thebois.roles is at 60% branch coverage:
image

I'll fix com.thebois.roles and try to get it to 100%. How should we handle com.thebois.world?

**Comment by jonathan-lindqvist** [comment]
> @jonathan-lindqvist yeah that's fair, it's a huge PR. Part of the problem is that I needed to pull in a lot of changes from the other branches, which meant I had to modify everything I pulled in to match the new systems.

Yes, I agree to a certain degree that this was sort of inevetable! 
 
> The other part is that implementing the InfoView and the new way of rendering views actually _is_ a part of the user story #16. Having user stories "for the programmers" like "As a programmer I want to refactor the view rendering so that there can be UI elements." isn't correct iirc. Since the user stories are vertical slices of everything that needs to be implemented to finish the story, the rendering refactor is supposed to be baked into the user story.

My bad, yeah you are right we shouldn't have user stories that are for programmers, but we can have refactor tasks, seperate from user stories. 

And it being a part of the user story is sort of correct. Displaying it to the user is a task, but the task needs our rendering to be refactored since we couldn't use shapes. And since its such a big refactor it should be a seperate task. 

Here the whole team failed, we should have seen that when we couldn't use shape for rendering we should have recognized it as "a problem". 

> 
> On the other hand, I agree that it's still a good idea to split out these "user story agnostic"-implementations into separate PRs. We could do those in branches separate from the user story branch that actually needs them, to get the benefits you mentioned. Then the problem would be linking branches to their corresponding user stories, but the benefits probably outweigh the downsides.

As mentioned now we can just link it to a task! 👍 

> 
> Looking at it from a different perspective though, no matter how much we split up this PR into multiple different PRs from multiple branches, you'd still have to review all of it. You'd just be forced to split it up over a few days. For this PR, you can still review it in parts and come back to it over the span of a few days until you're done. Doesn't necessarily have to be completed in a single session, imo.
> 
To take an example to the extreme we could finish the game in one branch and just review it at the end of the project. That wouldn't be great 😦

Now everything is connected (like it should be, its being used) and now I also have to figure out what connections are relevant for what I would be trying to review (if I split it up in multiple days). When doing all this its eaiser to slip up in the review procss and miss something, therefore its easier for bugs, etc to make it past.


> edit: I've added a discussion item to friday's protocol, so we can discuss how to handle these userstory-sub-branches!

👍

**Comment by jonathan-lindqvist** [comment]
> I'll fix `com.thebois.roles` and try to get it to 100%. How should we handle `com.thebois.world`?
A task for fixing test? So somone can focus 100% on tests for all classes in the model

**Comment by martinjonsson01** [comment]
>And since its such a big refactor it should be a seperate task.

>we can just link [the refactoring branch] to a task

I agree, that sounds like a good way of splitting it.



>To take an example to the extreme we could finish the game in one branch and just review it at the end of the project. That wouldn't be great 😦

Fair enough! 😆

>Now everything is connected (like it should be, its being used) and now I also have to figure out what connections are relevant for what I would be trying to review (if I split it up in multiple days). When doing all this its eaiser to slip up in the review procss and miss something, therefore its easier for bugs, etc to make it past.

I see, yeah, that makes it quite a lot harder.

>A task for fixing test? So somone can focus 100% on tests for all classes in the model

A task is part of a user story, though. Otherwise we're back to the user story being "because the programmers want 100% coverage". So I suppose whomever is working with the world right now will have to take that task as part of their story. That is, unless I should just do it now and get it over with? 😛

File: src/main/java/com/thebois/models/beings/Pawn.java (hunk 6 above)
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/beings/Pawn.java:13]: > 
> 
> "/* Temporary hard-coded world size. Should be removed when pathfinding is implemented. */"

Aaaaaaaaaaaaah, I see. Then it's sort of OK, it's pretty much a `TODO: ` comment.

> 
> It was either that, or send a reference to the `World` instance to `Pawn` which is even worse.

That I agree 100% with! 👍 

> 
> That's a nice idea, we could place it as a `public static` constant in `World`. Problem with that is that is global data, which is still a huge code smell.

Hmm, true.

> 
> My reasoning was that no matter how we implement this right now, it's too much work for something that'll be removed right after this PR is merged and I begin work on #13 😛

As long as we don't forget it, so it because a bug in the future 😄

**Review by Matte02 — COMMENTED** [review]
Overall very well done!

Spent some time understanding and organizing myself in the code and these were some of the things I found. I will reread it again tomorrow and might find something more. 

Really nice with the `ParameterizedTest`, will add it to my branch as well. 

As @jonathan-lindqvist this is a massive PR and is a bit overwhelming to look at and review. Talking about how we deal with these giant PRs on the next meeting is a good idea.

File: src/main/java/com/thebois/models/beings/Pawn.java (hunk 6 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/Pawn.java:13]: >it's pretty much a TODO: comment.

Yeah... Which hurts to merge into master, but it's either that or have ugly dependencies, or just completely remove it and have pawns that don't move.

>As long as we don't forget it, so it because a bug in the future 😄

image

**Comment by Matte02** [comment]
> A task is part of a user story, though. Otherwise we're back to the user story being "because the programmers want 100% coverage". So I suppose whomever is working with the world right now will have to take that task as part of their story. That is, unless I should just do it now and get it over with? 😛

Either you do it now, or leave it for me when I merge mine with master.

File: src/main/java/com/thebois/views/IActorView.java (hunk 13 above)
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/views/IActorView.java:15]: Maybe `getWidgetOwner` or somthing similar?

File: src/main/java/com/thebois/models/beings/roles/BuilderRole.java [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+
+/**
+ * The builder constructs buildings.
+ */
+class BuilderRole extends AbstractRole {
+
+    @Override
+    public RoleType getType() {
+        return RoleType.BUILDER;
+    }
```
[the remark this answers is not in the document]
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/roles/BuilderRole.java:11]: I feel like having it in a method rather than a constructor is preferable as, generally speaking, methods are way more flexible than constructors. For example, you can't call the constructor multiple times. Therefore you can't ever change the `RoleType`. With a method, you can allow it to change dynamically. 

We probably don't want that, but it's just an example of the freedom we get by using methods rather than constructors. Is there a reason why constructors are better in this case?

File: src/main/java/com/thebois/models/beings/AbstractBeingGroup.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/AbstractBeingGroup.java:33]: >Wouldn't it be possible to use Iterable here instead?

Turning my own ways against me, I see. 😆
Yeah, no:
image


>And is it fine that we give a reference to the real beings and not copies of them as well? Like we do with Terrain and Structure. Since all the changes made to beings should go through BeingGroup?

Problem is that `Colony`, a subclass to `AbstractBeingGroup` needs to mutate the `IBeing`s. Since it's better to have subtypes access fields in their supertypes through a getter, `Colony` has to use `getBeings()`. Since `IBeing` represents an Entity, it should be okay since they are supposed to be mutable. There is only supposed to be a single canonical representation of a pawn in the game, for example. 

Question is, do we think `Terrain` and `Structure`s are Entities as well, or are they data objects? I'm leaning towards data objects for `Terrain` but Entity for `Structure, since terrain is just a thing while a structure is much more actively interacting with the world.

File: src/main/java/com/thebois/models/beings/roles/BuilderRole.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (Matte02) [reply src/main/java/com/thebois/models/beings/roles/BuilderRole.java:11]: Nothing really, just felt weird looking at 6 classes all having the same method, and the only difference being what comes after RoleType. 

And constructors are usually often similar to each other and sometimes only differ in what arguments are sent in when they are subtypes. 

You are right though with saying methods are usually more flexible, so it might just be better to go with that and I don't mind either.

File: src/main/java/com/thebois/models/beings/roles/RoleFactory.java [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    /**
+     * Creates an instance of every type of role.
+     *
+     * @return All roles
+     */
+    public static Collection<AbstractRole> all() {
+        final ArrayList<AbstractRole> roles = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/roles/RoleFactory.java:37]:
```java
public static Collection<AbstractRole> all() {
  final Collection<AbstractRole> roles = new ArrayList<>();
  return roles;
```
is equivalent to
```java
public static Collection<AbstractRole> all() {
  final ArrayList<AbstractRole> roles = new ArrayList<>();
  return roles;
```
because in both cases the static type of the returned object is `Collection<AbstractRole>` and the dynamic type is `ArrayList<AbstractRole>`. 

It's just a matter of style. The benefit of keeping it as an `ArrayList` until it is returned is that the method has internal access to more methods to operate on the list. Which I think is good, right?

**Comment by martinjonsson01** [comment]
>Either you do it now, or leave it for me when I merge mine with master.

I'm getting tired of writing these tests so I'll leave it to you @Matte02 if it's not too much trouble! 😅

File: src/main/java/com/thebois/models/beings/roles/BuilderRole.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/roles/BuilderRole.java:11]: Until someone comes up with a good reason to prefer constructors, I'd prefer us to go with methods for the above reason.

File: src/main/java/com/thebois/views/IActorView.java (hunk 13 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/views/IActorView.java:15]: in that case it's more of a widget container, it doesn't really own them, just positions them in a hierarchy with the top node being the "root" node. Hence the name `getRoot`. Would `getWidgetContainer` be clearer?

File: src/main/java/com/thebois/models/beings/IBeing.java [hunk 22]
```diff
@@ -1,19 +1,35 @@
...
 
+import com.thebois.models.IDeepClonable;
 import com.thebois.models.Position;
+import com.thebois.models.beings.roles.AbstractRole;
 
 /**
  * An entity.
  */
-public interface IBeing {
+public interface IBeing extends IDeepClonable<IBeing> {
```
Comment (JacobBredin) [inline src/main/java/com/thebois/models/beings/IBeing.java]: Should IBeing really extend IDeepClonable? If AbstractBeing shouldn't be of that type, why should IBeing be? Am I missing something?

Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/IBeing.java]: Nope, you're absolutely right. Have removed it already, will be pushing it up soon.

Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/IBeing.java]: Or actually, it's already pushed. Did you make this comment a few hours ago maybe? 😛

Comment (JacobBredin) [reply src/main/java/com/thebois/models/beings/IBeing.java]: what. ok I'm confused.

File: src/main/java/com/thebois/models/beings/IBeing.java (hunk 5 above)
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/IBeing.java:14]: Doesn't seem to be an option in checkstyle. Maybe with some regex magic, idk.

Anyways, added it as a discussion point for Friday.

**Comment by jonathan-lindqvist** [comment]
> A task is part of a user story, though. Otherwise we're back to the user story being "because the programmers want 100% coverage". So I suppose whomever is working with the world right now will have to take that task as part of their story. That is, unless I should just do it now and get it over with? 😛

User stories needs task, tasks don't need user stories. We use user stories to prioritize what to work on and relate it to the customer to provide them with value. User stories provides the value to the customer, not tasks. Tasks are used to define what is need to be done, and when a task is connected to a user story we can say that that user story is done.

For examples the scrumboards we saw as an example were filled with tasks, that may or may not have been related to a user story. If it is related to a user story it _should_ be prioritized higher.

**Comment by martinjonsson01** [comment]
>User stories needs task, tasks don't need user stories.

I see. So that would mean we are free to create whatever tasks we want to, even if it isn't something that really benefits our users.

>If it is related to a user story it should be prioritized higher.

Don't we always have tasks tied to user stories, though? That would mean we never work on the standalone-tasks. 

And is there such a thing as a standalone-task? Isn't everything we do tied somehow to the final game, which the user cares about, which we specify through user stories? Seems like there's always a user story to tie a task to, and if there's not, then the task probably contains something we don't need to do.

**Review by Matte02 — APPROVED** [review]
Only 1 Issue left on Proof of Concept then! Almost done.

**Review by jonathan-lindqvist — APPROVED** [review]
Hope you read this before you merge. I trust that you fix my comment before you merge, just approving since it's unnecessary for me to review that you removed the null comparison and that would block you :P. 

So when that is fixed you can just merge it! Great job 👍

File: src/main/java/com/thebois/models/beings/Colony.java (hunk 4 above)
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/beings/Colony.java]: For now its OK, if we add more stuff and Colony actually does something we will need to look into a solution.

**Comment by jonathan-lindqvist** [comment]
> Don't we always have tasks tied to user stories, though? That would mean we never work on the standalone-tasks.
> 
> And is there such a thing as a standalone-task? Isn't everything we do tied somehow to the final game, which the user cares about, which we specify through user stories? Seems like there's always a user story to tie a task to, and if there's not, then the task probably contains something we don't need to do.

For example tests, it doesn't provide the user with immediate value but it does for us. So when someone is blocked (waiting for something to be implemented) or finished all work on user stories for the sprint, then they can take the low priority (standalone) tasks.

**Comment by martinjonsson01** [comment]
>For example tests

Alright, that kind of makes sense. Although, we should strive to include all necessary tests in the user stories themselves. It's possible that sometimes tests are missed and have to be done separately. 👍

File: src/main/java/com/thebois/models/beings/AbstractBeing.java [hunk 23]
```diff
@@ -1,42 +1,75 @@
...
+    }
+
+    @Override
+    public Position getPosition() {
+        return currentPosition.deepClone();
+    }
+
+    @Override
+    public AbstractRole getRole() {
+        if (role == null) return null;
```
[the remark this answers is not in the document]
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/AbstractBeing.java:61]: Nice catch, I did a search for all `null`s in the project, but must've missed this one!

[the remark this answers is not in the document]
Comment (martinjonsson01) [reply src/main/java/com/thebois/models/beings/AbstractBeing.java:61]: Since we never should use null as a role, I'm removing the null check entirely so we can see the errors whenever someone accidentally does set a role to null.
